openslides_backend/action/actions/user/assign_meetings.py

```diff
@@ -0,0 +1,124 @@
+        group_name = instance.pop("group_name")
+
+        user = self.datastore.get(
+            fqid_from_collection_and_id("user", user_id),
+            [
+                "meeting_ids",
+                "group_$_ids",
+                *[f"group_${meeting_id}_ids" for meeting_id in meeting_ids],
+            ],
+        )
+        user_meeting_ids = set(user.get("meeting_ids", []))
+        filter_ = And(
+            FilterOperator("name", "=", group_name),
+            Or(
+                *[
+                    FilterOperator("meeting_id", "=", meeting_id)
+                    for meeting_id in meeting_ids
+                ]
+            ),
+        )
+        groups = self.datastore.filter("group", filter_, ["meeting_id", "user_ids"])
+        groups_meeting_ids = set(group["meeting_id"] for group in groups.values())
+        # Now split the meetings in the 3 categories
+        self.success = groups_meeting_ids
+        success_update = self.success.difference(user_meeting_ids)
+        self.standard_meeting_ids = meeting_ids.difference(
+            groups_meeting_ids
+        ).difference(user_meeting_ids)
+        self.nothing_meeting_ids = meeting_ids.difference(
+            groups_meeting_ids
+        ).intersection(user_meeting_ids)
+        # check if all meetings are set into one category
+        assert (
+            self.success.union(self.standard_meeting_ids).union(
+                self.nothing_meeting_ids
+            )
+            == meeting_ids
+        )
+
+        if not success_update:
+            raise ActionException(
+                f"Don't find a group with groupname {group_name} to assign to in any meeting."
+            )
+
+        # fill the instance for the update
+        if success_update.union(self.standard_meeting_ids):
+            instance["group_$_ids"] = {}
+        for meeting_id in success_update:
+            instance["group_$_ids"][meeting_id] = list(
+                set(user.get(f"group_${meeting_id}_ids") or []).union(
+                    set(self.find_group_id(meeting_id, groups))
+                )
+            )
+        meetings = {}
+        if self.standard_meeting_ids:
+            get_many_request = GetManyRequest(
+                "meeting", list(self.standard_meeting_ids), ["default_group_id"]
+            )
+            get_many_result = self.datastore.get_many([get_many_request])
+            meetings = get_many_result.get("meeting", {})
+        for meeting_id in self.standard_meeting_ids:
+            meeting = meetings[meeting_id]
+            instance["group_$_ids"][meeting_id] = [meeting["default_group_id"]]
+
+        return instance
+
+    def find_group_id(self, meeting_id: int, groups: Dict[int, Any]) -> List[int]:
+        for group_id, group in groups.items():
+            if group.get("meeting_id") == meeting_id:
+                return [group_id]
+        raise ActionException("Error, could not find group-id")
```

Comment:
If you transfer the result first into a map meeting_id<->group_id, you can directly access the correct group by the meeting id and do not have to iterate over all groups every time. Also, why return a list which always contains only one element?

Reply:
Replace `find_group_id` with the meeting_to_group map. Use `set([group_id])` there to init a one element set.